src/storage/store/proposal.rs

```diff
@@ -0,0 +1,97 @@
+use crate::proto::{Block, FullProposal, Height};
+use crate::storage::constants::RootPrefix;
+use crate::storage::db::{RocksDB, RocksdbError};
+use informalsystems_malachitebft_core_types::Round;
+use prost::Message;
+use std::sync::Arc;
+use thiserror::Error;
+use tracing::error;
+
+#[derive(Error, Debug)]
+pub enum ProposalStorageError {
+    #[error(transparent)]
+    RocksdbError(#[from] RocksdbError),
+
+    #[error("Block missing header")]
+    BlockMissingHeader,
+
+    #[error("Block missing height")]
+    BlockMissingHeight,
+
+    #[error("Too many blocks in result")]
+    TooManyBlocksInResult,
+
+    #[error("Error decoding shard chunk")]
+    DecodeError(#[from] prost::DecodeError),
+}
+
+/** A page of messages returned from various APIs */
+pub struct BlockPage {
+    pub blocks: Vec<Block>,
+    pub next_page_token: Option<Vec<u8>>,
+}
+
+fn make_proposal_key(height: u64, round: i64) -> Vec<u8> {
+    // Store the prefix in the first byte so there's no overlap across different stores
+    let mut key = vec![RootPrefix::Block as u8];
+    // Store the block number in the next 8 bytes
+    key.extend_from_slice(&height.to_be_bytes());
+    key.extend_from_slice(&round.to_be_bytes());
+
+    key
+}
+
+pub fn put_proposal(db: &RocksDB, proposal: FullProposal) -> Result<(), ProposalStorageError> {
```

Comment:
These functions can go into the structs impls, no need to keep them separate. The hub used to do this, but I think it's not as clean. We should start fixing it.